File: src/engine/adapters/clickhouse/definition.ts
import type {
  ConstraintMeta,
  DefinitionSection,
  ObjectDefinition,
} from '../../../shared/domain/definition';
import { encodePath, type NodePath } from '../../../shared/domain/tree';
import { AdapterError } from '../errors';
import { stripOneTrailingSemicolon } from '../sql-text';
import type { QueryExecutor } from './catalog';
import * as catalog from './catalog';

// D18: system.constraints has no primary/unique/foreign-key rows at all (F17/F16 — a MergeTree
// PRIMARY KEY is a sparse index, not a constraint the catalog tracks this way), so this section is
// CHECK-only rather than a partial imitation of what postgres/mariadb's constraints list shows.
function toConstraintMeta(row: { name: string; expression: string }): ConstraintMeta {
  return { name: row.name, type: 'check', definition: row.expression };
}

// P23 D6: the "Table properties" section — this adapter's first real use of the generic
// name/value block every other SQL engine's definition() leaves empty, since none of Engine,
// Sorting key, Partition key or the sparse Primary key expression has anywhere else in
// ObjectDefinition to live.
function buildTableSection(target: catalog.ReadTarget): DefinitionSection {
  const rows: DefinitionSection['rows'] = [
    { name: 'Engine', value: target.engine, detail: null },
    { name: 'Sorting key', value: target.sortingKey || '(none)', detail: null },
    { name: 'Primary key', value: target.primaryKeyExpression || '(none)', detail: null },
    { name: 'Partition key', value: target.partitionKey || '(none)', detail: null },
    {
      name: 'Total rows',
      value: target.totalRows !== null ? target.totalRows.toLocaleString() : '(unknown)',
      detail: null,
    },
  ];
  if (target.comment) rows.push({ name: 'Comment', value: target.comment, detail: null });
  return { title: 'Table properties', rows };
}

export async function buildDefinition(
  exec: QueryExecutor,
  segments: NodePath['segments'],
  schema: string,
  object: { kind: 'table' | 'view' | 'matview'; name: string },
): Promise<ObjectDefinition> {
  const target = await catalog.getReadTarget(exec, schema, object.name);
  const statement = stripOneTrailingSemicolon(target.createTableQuery);
  if (statement === '') {
    throw new AdapterError('E_QUERY', `no definition returned for "${schema}"."${object.name}"`);
  }

  const checkRows = catalog.listCheckConstraints(target.createTableQuery);
  const constraints = checkRows.map(toConstraintMeta);

  const notes = [
    'A MergeTree PRIMARY KEY is a sparse index, not a uniqueness constraint — see Table properties.',
    'ClickHouse has no foreign keys.',
  ];

  return {
    path: encodePath(segments),
    kind: object.kind,
    qualifiedName: `${schema}.${object.name}`,
    statements: [statement],
    language: 'sql',
    origin: 'server',
    notes,
    constraints,
    documentSchema: null,
    sections: [buildTableSection(target)],
    generatedAt: new Date().toISOString(),
  };
}

File: apps/kira-studio/internal/adapters/clickhouse/definition.go
package clickhouse

import (
	"context"
	"strconv"
	"time"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/adapters"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/storage/model"
)

func toConstraintMeta(row constraintRow) model.ConstraintMeta {
	return model.ConstraintMeta{Name: row.Name, Type: "check", Definition: row.Expression}
}

// buildTableSection is definition.ts's own — P23 D6: the "Table properties" section, this
// adapter's first real use of the generic name/value block every other SQL engine's definition()
// leaves empty, since none of Engine, Sorting key, Partition key or the sparse Primary key
// expression has anywhere else in ObjectDefinition to live.
func buildTableSection(target ReadTarget) model.DefinitionSection {
	sortingKey := target.SortingKey
	if sortingKey == "" {
		sortingKey = "(none)"
	}
	primaryKey := target.PrimaryKeyExpression
	if primaryKey == "" {
		primaryKey = "(none)"
	}
	partitionKey := target.PartitionKey
	if partitionKey == "" {
		partitionKey = "(none)"
	}
	totalRows := "(unknown)"
	if target.TotalRows != nil {
		totalRows = formatThousands(*target.TotalRows)
	}
	rows := []model.DefinitionSectionRow{
		{Name: "Engine", Value: target.Engine},
		{Name: "Sorting key", Value: sortingKey},
		{Name: "Primary key", Value: primaryKey},
		{Name: "Partition key", Value: partitionKey},
		{Name: "Total rows", Value: totalRows},
	}
	if target.Comment != nil && *target.Comment != "" {
		rows = append(rows, model.DefinitionSectionRow{Name: "Comment", Value: *target.Comment})
	}
	return model.DefinitionSection{Title: "Table properties", Rows: rows}
}

// formatThousands is toLocaleString()'s own thousands-grouping, the one piece of that call site
// worth porting rather than dropping: a plain strconv.FormatInt would silently lose the grouping
// dots/commas the TS's own definition view has always shown.
func formatThousands(n int64) string {
	s := strconv.FormatInt(n, 10)
	neg := ""
	if s[0] == '-' {
		neg, s = "-", s[1:]
	}
	var out []byte
	for i, c := range []byte(s) {
		if i > 0 && (len(s)-i)%3 == 0 {
			out = append(out, ',')
		}
		out = append(out, c)
	}
	return neg + string(out)
}

var definitionSupportedKinds = map[string]bool{"table": true, "view": true, "matview": true}

// buildDefinition is definition.ts's own.
func buildDefinition(ctx context.Context, h *Handle, queryID string, op *adapters.OpCtx, track TrackQuery, segments []model.PathSegment, schema string, kind, name string) (model.ObjectDefinition, error) {
	target, err := getReadTarget(ctx, h, queryID, op, track, schema, name)
	if err != nil {
		return model.ObjectDefinition{}, err
	}
	statement := adapters.StripOneTrailingSemicolon(target.CreateTableQuery)
	if statement == "" {
		return model.ObjectDefinition{}, adapters.New(adapters.CodeQuery, `no definition returned for "`+schema+`"."`+name+`"`, nil)
	}

	checkRows := listCheckConstraints(target.CreateTableQuery)
	constraints := make([]model.ConstraintMeta, len(checkRows))
	for i, r := range checkRows {
		constraints[i] = toConstraintMeta(r)
	}

	notes := []string{
		"A MergeTree PRIMARY KEY is a sparse index, not a uniqueness constraint — see Table properties.",
		"ClickHouse has no foreign keys.",
	}

	return model.ObjectDefinition{
		Path: model.EncodePath(segments), Kind: kind, QualifiedName: schema + "." + name,
		Statements: []string{statement}, Language: "sql", Origin: "server", Notes: notes,
		Constraints: constraints, DocumentSchema: nil, Sections: []model.DefinitionSection{buildTableSection(target)},
		GeneratedAt: time.Now().UTC().Format(time.RFC3339Nano),
	}, nil
}
